pkg/globalnet/controllers/global_ingressip_controller.go

```diff
@@ -0,0 +1,159 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/submariner-io/admiral/pkg/federate"
+	"github.com/submariner-io/admiral/pkg/syncer"
+	"github.com/submariner-io/admiral/pkg/util"
+	submarinerv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/ipam"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+func NewGlobalIngressIPController(config syncer.ResourceSyncerConfig, pool *ipam.IPPool) (Interface, error) {
+	var err error
+
+	klog.Info("Creating GlobalIngressIP controller")
+
+	controller := &globalIngressIPController{
+		baseIPAllocationController: newBaseIPAllocationController(pool),
+	}
+
+	_, gvr, err := util.ToUnstructuredResource(&submarinerv1.GlobalIngressIP{}, config.RestMapper)
+	if err != nil {
+		return nil, err
+	}
+
+	client := config.SourceClient.Resource(*gvr).Namespace(corev1.NamespaceAll)
+	list, err := client.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	federator := federate.NewUpdateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll)
+
+	for i := range list.Items {
+		err = controller.reserveAllocatedIPs(federator, &list.Items[i])
```

Comment:
Just for info: Along with reserving the IPs we also have to sync the IPTable rules, hence in my PR [https://github.com/submariner-io/submariner/pull/1387], I removed this code and included similar logic in OnCreate method. We can leave it this way, when I work on the IPTables stuff, I'll do the necessary changes.

Reply:
This was done on startup so we can ensure each object will be able to reserve its previously allocated IPs (unless the global CIDR changed). Otherwise, in doing this on callbacks, we may process a newly created object first which could allocate a previously used block, in which case we'd need to reallocate.